### ISSUE TITLE
Better timeout support: `.ftp-deploy-sync-state.json` model / early generation, number of files / folders to sync...

### ISSUE THREAD

**Issue by DamChtlv** [SamKirkland/ftp-deploy#34]
I'm having an issue where after some time creating folders / files, the action stop working _(seems to freeze)_ and finally resolve to this: `Error: Timeout (control socket)`

**Log representation:**  

```
----------------------------------------------------------------
🚀 Thanks for using ftp-deploy. Let's deploy some stuff!   
----------------------------------------------------------------
If you found this project helpful, please support it
by giving it a ⭐ on Github --> https://github.com/SamKirkland/FTP-Deploy-Action
or add a badge 🏷️ to your projects readme --> https://github.com/SamKirkland/FTP-Deploy-Action#badge
----------------------------------------------------------------
No file exists on the server "./autodeploytest/.ftp-deploy-sync-state.json" - this must be your first publish! 🎉
The first publish will take a while... but once the initial sync is done only differences are published!
If you get this message and its NOT your first publish, something is wrong.
----------------------------------------------------------------
Local Files:	340
Server Files:	0
----------------------------------------------------------------
Calculating differences between client & server
----------------------------------------------------------------

📁 Create: ... (a lot)
📄 Upload: ... (a lot)

----------------------------------------------------------------
Making changes to 340 files/folders to sync server state
Uploading: 4.97 MB -- Deleting: 0 B -- Replacing: 0 B
----------------------------------------------------------------

creating folder "..." (some)
uploading "..." (some)

...then at one point, it freeze here i don't know why...
and after the specific timeout, it finally resolve to this:

----------------------------------------------------------------
--------------  🔥🔥🔥 an error occurred  🔥🔥🔥  --------------
----------------------------------------------------------------

----------------------------------------------------------------
----------------------  full error below  ----------------------
----------------------------------------------------------------

Error: Timeout (control socket)
    at Socket.<anonymous> (/home/runner/work/_actions/SamKirkland/FTP-Deploy-Action/v4.3.4/dist/index.js:5288:33)
    at Object.onceWrapper (node:events:627:28)
    at Socket.emit (node:events:513:28)
    at Socket._onTimeout (node:net:550:8)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
Error: Error: Timeout (control socket)
```

The problem is that it starts over from the beginning if i trigger it again.  
I tried changing default `timeout` value from `120000` to `15000`, but it always result in timeout error.

It would be nice to have an option to create `.ftp-deploy-sync-state.json` before a timeout occurs so it can actually make progress _(like one second before set timeout)_  
Or to be able to set a number of files / folders to sync so it stops after this number and generate `.ftp-deploy-sync-state.json` to store progress.
Or even have a `.ftp-deploy-sync-state.json` model / template, so we can see how it is structured and create it ourselves on the server so it can resume where it stopped previously

**Comment by markparnell** [comment]
Having a very similar issue here - it runs for around 20 minutes and successfully uploads several hundred files and folders before returning an error:

```
----------------------------------------------------------------
----------------------  full error below  ----------------------
----------------------------------------------------------------

FTPError: 425 Unable to build data connection: Operation not permitted
    at FTPContext._onControlSocketData (/home/runner/work/_actions/SamKirkland/FTP-Deploy-Action/v4.3.4/dist/index.js:5252:39)
    at TLSSocket.<anonymous> (/home/runner/work/_actions/SamKirkland/FTP-Deploy-Action/v4.3.4/dist/index.js:5096:44)
    at TLSSocket.emit (node:events:513:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:285:11)
    at TLSSocket.Readable.push (node:internal/streams/readable:228:10)
    at TLSWrap.onStreamRead (node:internal/stream_base_commons:190:23) {
  code: 425
}
Error: FTPError: 425 Unable to build data connection: Operation not permitted
```

Having the ability for it to update the state after every _N_ files or something would be really handy. In our case it would also be helpful if we could flag what commit the remote is up to, as we are migrating from a different system and don't need it to upload anything initially, but that's a separate issue.

**Comment by mandrasch** [comment]
+1 

```
timeout: 60000
protocol: 'ftps'
dangerous-clean-slate: false
```
```
Sun, 15 Mar 2026 09:36:30 GMTLast published on 📅 Wednesday, March 11, 2026 at 9:58 PM
Sun, 15 Mar 2026 09:36:30 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:36:30 GMTLocal Files:	346
Sun, 15 Mar 2026 09:36:30 GMTServer Files:	299
Sun, 15 Mar 2026 09:36:30 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:36:30 GMTCalculating differences between client & server
Sun, 15 Mar 2026 09:36:30 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:36:30 GMT📁 Create: blog/kategorie

...... after some time


Sun, 15 Mar 2026 09:36:56 GMTuploading "blog/kategorie/aktueller-stand/index.html"
Sun, 15 Mar 2026 09:37:56 GMT
Sun, 15 Mar 2026 09:37:56 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:37:56 GMT--------------  🔥🔥🔥 an error occurred  🔥🔥🔥  --------------
Sun, 15 Mar 2026 09:37:56 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:37:56 GMT
Sun, 15 Mar 2026 09:37:56 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:37:56 GMT----------------------  full error below  ----------------------
Sun, 15 Mar 2026 09:37:56 GMT----------------------------------------------------------------
Sun, 15 Mar 2026 09:37:56 GMT
Sun, 15 Mar 2026 09:37:56 GMTError: Timeout when trying to open data connection to XX.XX.XX.XX:9347
Sun, 15 Mar 2026 09:37:56 GMT    at Socket.handleTimeout (/home/runner/work/_actions/SamKirkland/FTP-Deploy-Action/v4.3.6/dist/index.js:6256:20)
Sun, 15 Mar 2026 09:37:56 GMT    at Socket.emit (node:events:524:28)
Sun, 15 Mar 2026 09:37:56 GMT    at Socket._onTimeout (node:net:595:8)
Sun, 15 Mar 2026 09:37:56 GMT    at listOnTimeout (node:internal/timers:581:17)
Sun, 15 Mar 2026 09:37:56 GMT    at process.processTimers (node:internal/timers:519:7)
Sun, 15 Mar 2026 09:37:56 GMTError: Error: Timeout when trying to open data connection to XXX...:9347
```

Asked my german webhosting provider, they asked if there are many sftp connections opened in parallel which might cause trouble. But this was jus their educated guess. They suggested using rsync/ssh because of stability, but I'd love to keep SFTP because I can restrict access to specific folders. 🤷 And imho a lot of cheap webhosting providers only offer SFTP, therefore I like this action because it can be used by everyone. Thanks for providing this action!
